### ISSUE TITLE
Model produces three different outcomes for the same input depending on how I measure it

### ISSUE THREAD

**Issue by ichitaka** [tensorflow/tensorflow#43164]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.15.6
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.3.0
- Python version: 3.7.7

**Describe the current behavior**
When I take the loss from the history object or the result from the evaluate function on the training data or from a numpy calculation of mean squared error I receive different results.
This is an output of my code below:
```
4.520166488224531 
4.409448146820068 
4.139582633972168
```

**Describe the expected behavior**
I would receive the same results for all different evaluation calculations.
Like:
```
4.520166488224531 
4.520166488224531 
4.520166488224531 
```
**Standalone code to reproduce the issue**

You can find a colab notebook with the issue here: 
https://colab.research.google.com/gist/ichitaka/c97ce36d57184ba154c8840802a8e01d

```
from tensorflow.keras.layers import Input, Dense, Activation
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import RMSprop
import numpy as np
x = np.random.normal(1, 2, 10)
y = np.random.normal(1, 2, 10)

model = Sequential()
model.add(Input(1))
model.add(Dense(10))
model.add(Activation("sigmoid"))
model.add(Dense(1))
model.summary()


rmsprop = RMSprop(lr=0.01)
model.compile(loss='mean_squared_error', optimizer=rmsprop)
hist = model.fit(x=x,y=y)
_ = model.evaluate(x, y, verbose=0)
pred = model.predict(x)
print(np.mean((y-pred)**2), '\n', hist.history['loss'][0], '\n', _)
```

**Comment by ichitaka** [comment]
So seems like the numpy function had issues with the dimensionality. When I fix that, evaluate and the numpy function return the same. Also adding the training data as validation data returns also the same value as the other two. So it seems, like the loss in history to be a value before the last gradient is applied. So I would consider this to not be a bug, just unexpected behaviour.
